crates/shared/src/sources/balancer_v2/pool_init.rs

```diff
@@ -20,23 +22,36 @@ pub trait PoolInitializing: Send + Sync {
 ///
 /// This can be used to index all pools from events instead of relying on the
 /// Balancer subgraph for example.
-pub struct EmptyPoolInitializer(u64);
+pub struct EmptyPoolInitializer {
+    chain_id: u64,
+    web3: Web3,
+}
 
 impl EmptyPoolInitializer {
     /// Creates a new empty pool initializer for the specified chain ID.
     #[cfg(test)]
-    pub fn for_chain(chain_id: u64) -> Self {
-        Self(chain_id)
+    pub fn for_chain(chain_id: u64, web3: Web3) -> Self {
+        Self { chain_id, web3 }
     }
 }
 
 #[async_trait::async_trait]
 impl PoolInitializing for EmptyPoolInitializer {
     async fn initialize_pools(&self) -> Result<RegisteredPools> {
         let fetched_block_number =
-            deployment_block(BalancerV2Vault::raw_contract(), self.0).await?;
+            deployment_block(BalancerV2Vault::raw_contract(), self.chain_id).await?;
```

Comment:
Maybe as a future improvement, we can include the block number and hash as part of the deployment information (and not just block number).